### PR TITLE
DEVEXP-131: Configure constraining column value storage in MarkLogic

### PR DESCRIPTION
**Pull request by BillFarber**
DEVEXP-131: Configure constraining column value storage in MarkLogic

### CODE REVIEW

File: src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java
Comment (rjrudin) [inline src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java:11]: I think these two tests can be merged, as there's really just one scenario:

1. Load the fifteen authors
2. Run the Task
3. Verify that the constraint value document was written correctly (as you're doing already)
4. Load 3 more authors - two with IDs higher than the highest one that was captured, and one with an ID that is lowest that the highest one
5. Run the Task again
6. Verify that only the two new authors with a higher ID are returned
7. And then verify that the constraint value doc was updated as well

Comment (rjrudin) [reply src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java:11]: Also, I think it's worth adding an error scenario too - i.e. do something to cause the read of the JSON doc to fail, such as first writing some random XML to that URI. Then load the 15 authors, and run the Task, and verify that you get null back since the read of the JSON doc failed.

**Review by rjrudin — CHANGES_REQUESTED** [review]
This looks good, and while I haven't QA'ed it yet, looks functionally correct.

File: src/main/java/com/marklogic/kafka/connect/source/MarkLogicConstraintValueStore.java
Comment (rjrudin) [inline src/main/java/com/marklogic/kafka/connect/source/MarkLogicConstraintValueStore.java:43]: This use case of - build up a doc to be written to ML, based on some String properties - seems like another that the Java Client should support. We actually have a class to do this already in the Kafka project - `DocumentWriteOperationBuilder` - which I think you can reuse here, but don't worry about it in this PR - it's a refactor for later, and I want to think more about whether the Java Client should really be providing this functionality because we know it's a common use case (e.g. the NiFi connector does this exact same thing too).

**Review by rjrudin — COMMENTED** [review]
This looks good, I'll do some quick QA on it this morning and approve it after that.

File: src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java
Comment (rjrudin) [inline src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java:78]: Typo, should be "20" - though I would just do "all" to avoid the chance of the number being incorrect

File: src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java
Comment (rjrudin) [inline src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java:81]: This should be private, just to make it clear it's not used anywhere else. Also, this is going to fail not because the data is invalid, but because the doc is lacking a read permission and the user won't see it. I think it'd be better, and less code, to just reuse the ConstraintValueStore but pass in a value that isn't valid. 

Actually - I just tried this, and it appears to be a different scenario. I tried this:

    constraintValueStore.storeConstraintState("invalid-value");

And the behavior I got instead was that no rows were returned. 

That seems unexpected to me - i.e. I would expect that if there's any problem with the constraint value document (it's missing, it's invalid, or the doc is valid JSON but the value is invalid), I would get the same behavior - which I think should be that all rows are returned. 

I think that's a good scenario to add, and I think both that and this scenario of "Doc is invalid" should be in separate test methods. It makes this method easier to understand (it's 60 LOC, so a little long already), and makes it easier to understand what happens for these two error scenarios. 

A 4th error scenario - try writing the doc with no permissions. That should fail, and we want to document what happens there.

File: src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java
Comment (rjrudin) [inline src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java:14]: To be consistent with story/action-based test names, let's do "StoreConstraintValueInMarkLogicTest".

**Review by rjrudin — COMMENTED** [review]
Few comments, doing QA in CP now.

File: src/main/java/com/marklogic/kafka/connect/source/MarkLogicSourceConfig.java
Comment (rjrudin) [inline src/main/java/com/marklogic/kafka/connect/source/MarkLogicSourceConfig.java:25]: This is correct per the story, but the story is now a little out-of-date (no need to update it) since we landed on "constraintColumn" above. So for consistency, these should all be "constraintColumn" too.

File: src/main/java/com/marklogic/kafka/connect/source/MarkLogicConstraintValueStore.java
Comment (rjrudin) [inline src/main/java/com/marklogic/kafka/connect/source/MarkLogicConstraintValueStore.java:42]: This should use `.withFormat(Format.JSON)` so that ML knows to create a JSON document as opposed to relying on the user-defined URI.

File: src/main/java/com/marklogic/kafka/connect/source/RowBatcherSourceTask.java
Comment (rjrudin) [inline src/main/java/com/marklogic/kafka/connect/source/RowBatcherSourceTask.java:100]: I did a test here where I forced a failure by setting the uri property to "hi.xml" - ML complains because it assumes I want an XML document. But it could fail for other reasons too, like missing an update permissions.

The error then gets logged, stating "job failed to complete successfully" - but the source records that were retrieved are still returned. That would be surprising to a user - if the job failed (though it didn't, the job succeeded), I think a user would expect no data to be returned. 

I think that suggests the following changes:

1. The try/catch should only be around this block of code for getting and storing the max value
2. The error message should be specific to that goal and should be a warning instead of an error since we're still going to return data - i.e. "Unable to update max constraint value, cause: " + e.getMessage()

**Review by rjrudin — CHANGES_REQUESTED** [review]

File: src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java
Comment (BillFarber) [reply src/test/java/com/marklogic/kafka/connect/source/ConstraintStorageTest.java]: I'm not real sure I've covered what you're looking for. Maybe we should take a look at it together.

**Review by rjrudin — APPROVED** [review]

**Pull request merged**
